engine/robot.py

```diff
@@ -360,98 +375,131 @@ def track_obstacle(self):
+        goal_threshold = 3
+        init_x = self.robot_state.state[0]
+        init_y = self.robot_state.state[1]
+        init_pos = (init_x, init_y)
+
+        # last position of the robot when it's still on the (line from its initial position to its goal positions)
+        last_pos_on_line = init_pos
         has_left_init_thresh = False
-        did_dist_to_goal_decreased = False
         has_traversed_boundary = False
-        curr_dist_to_goal = self.calculate_dist(self.robot_state.goal_location, (self.robot_state.x_pos, self.robot_state.y_pos))
+        curr_dist_to_goal = self.calculate_dist(
+            self.robot_state.goal_location, curr_pos)
+        init_dist_to_goal = curr_dist_to_goal
+        turn_to_avoid_obstacle = False
         while True:
-            if self.robot_state.phase == Phase.fault:  # fault has priority over obstacle avoidance
+            # fault has priority over obstacle avoidance
+            if self.robot_state.phase == Phase.FAULT:
                 return None
-            else:
-                dist_from_init = self.calculate_dist((self.robot_state.x_pos, self.robot_state.y_pos), (init_x, init_y))
-                if dist_from_init > (self.robot_state.init_threshold + self.robot_state.noise_margin):
-                    has_left_init_thresh = True
-                if curr_dist_to_goal < self.robot_state.goal_threshold:  # exits obstacle avoidance if robot close to goal
+            curr_pos = (self.robot_state.state[0], self.robot_state.state[1])
+            curr_dist_to_goal = self.calculate_dist(
+                self.robot_state.goal_location, curr_pos)
+
+            dist_from_init = self.calculate_dist(curr_pos, init_pos)
+            # threshold for checking theta matches for the lines
+            on_line_tolerance = 0.1
+            is_on_line = self.is_on_line(
+                self.robot_state.goal_location, curr_pos, on_line_tolerance)
+            new_dist_to_goal = self.calculate_dist(
+                self.robot_state.goal_location, curr_pos)
+            if dist_from_init > init_threshold:
+                has_left_init_thresh = True
+            # exits obstacle avoidance if robot close to goal
+            if curr_dist_to_goal < goal_threshold:
+                self.set_phase(self.robot_state.prev_phase)
+                return None
+            elif has_traversed_boundary:
+                self.set_phase(Phase.FAULT)  # cannot reach goal
+                return None
+            # bug 2
+            elif is_on_line and (new_dist_to_goal < init_dist_to_goal):
+                # need to tune, in meters or whatever the unit x and y are
+                # without this condition, the robot could move closer to the goal (on a slant, not directly towards the goal)
+                #   but still be on the line, making the robot keep exiting obstacle avoidance when its effectively in the same position as before
+                threshold_to_recalculate_on_line = 3
```

Comment:
like parameters? They're technically already variables. Will we ever want to change the threshold? If we only change it when initially tuning it and nowhere else, it might not make sense to make it function parameters, but I can change

Reply:
If the robot is doing boundary traversal and it only moves a little bit and it's "on the line" (within threshold), we would have to exit obstacle avoidance even though it only moved a little bit and the robot is still in the same location and the obstacle is still there. we don't want to do that